src/main/java/org/spongepowered/api/item/inventory/InventoryArchetypes.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.inventory;
+
+import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;
+
+/**
+ * An enumeration of {@link InventoryArchetype}s.
+ */
+public class InventoryArchetypes {
+
+    // Base Archetypes
+    // TODO how to start?
+    public static final InventoryArchetype SLOT = DummyObjectProvider.createFor(InventoryArchetype.class, "SLOT");
+    public static final InventoryArchetype MENU_ROW = DummyObjectProvider.createFor(InventoryArchetype.class, "MENU_ROW");
```

Comment:
> > IMO we should not use events (listeners on plugins) for this and use a callback system. BiPredicate.

> Yes because callbacks in an otherwise entirely event-driven framework is a good practice </sarcasm>

Well, we also have [`ExecuteCallback`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/text/action/ClickAction.java#L125) for `Text` based menus.

And yes IMO its a good idea because its easier to understand and maintain for plugin devs.

**Example:**
Callback:
````java
InventoryArchetype TEAM_SELECTION = InventoryArchetype.builder()
.with(InventoryArchetype.MENU_BUTTON.toBuilder().with(Keys.CALLBACK, (player, inventory) -> this.joinTeam(player, inventory, Team.RED)).build())
.with(InventoryArchetype.MENU_BUTTON.toBuilder().with(Keys.CALLBACK, (player, inventory) -> this.joinTeam(player, inventory, Team.BLUE)).build())
.register();

public void openTeamSelectionMenu(Player player) {
    TEAM_SELECTION.openFor(player);
}

public void joinTeam(Player player, Inventory inventory, Team team) {
    inventory.close();
    player.sendMessage("You joined team " + team.getName());
}
````
Event driven:
````java
InventoryArchetype TEAM_SELECTION = InventoryArchetype.builder()
.with(InventoryArchetype.MENU_BUTTON.toBuilder().with(Keys.CALLBACK_ID, "MYPLUGIN:joinRed").build())
.with(InventoryArchetype.MENU_BUTTON.toBuilder().with(Keys.CALLBACK_ID, "MYPLUGIN:joinBlue").build())
.register();

public void openTeamSelectionMenu(Player player) {
    TEAM_SELECTION.openFor(player);
}

public void joinTeam(Player player, Inventory inventory, Team team) {
    inventory.close();
    player.sendMessage("You joined team " + team.getName());
}

public void onMenuClick(InventoryClickEvent.ClickMenuOption event) {
    Player player = event.getPlayer();
    Inventory inventory = event.getInventory();
    String callbackId = event.foo().get(Keys.CALLBACK_ID);
    switch(callbackId) {
        case "MYPLUGIN:joinRed": 
            joinTeam(player, inventory, Team.RED);
            return;
        case "MYPLUGIN:joinBlue": 
            joinTeam(player, inventory, Team.BLUE);
            return;
        case "MYPLUGIN:leave": 
            [...]
            return;
        case "MYPLUGIN:invite": 
            [...]
            return;
        [...]
    }
}
````

**Possible issues**
Callback:
* (We use events otherwise)
* ???

Event driven:
* Possible conflicts with unqualified callback ids
* The button definition is separated from its action
* More event listeners for those menu events
* The listener method looks ugly/messy since it has to contain all cases, that could appear in the plugin.
* More code for plugins to write -> less readable

**Server implementation effort**
Should be mostly the same for both versions, since you have to check whether you are in a special control menu in both cases.
You have slight overhead in the callback version due to the necessary mapping (But that is basically a map lookup).

--------------------------------------

> > The event handler are just overhead that only have to exist once in the impl with this approach.

> The who-what with the where now?

The event handlers in plugins are overhead, because you could implement a single one in the server impl that does the necessary `button -> action` mapping.
See my above code examples:
The first one does not need a `onMenuClick` event handler since the server does that internally.
In the second example every plugin has to register a `onMenuClick` event handler to deal with its button itself.
Both are able to do exactly the same actions, just with other syntax and code placement.